src/rust/engine/process_execution/src/children.rs

```diff
@@ -47,14 +51,62 @@ impl ManagedChild {
     })
   }
 
-  /// Kill the process's unique PGID or return an error if we don't have a PID or cannot kill.
-  pub fn kill_pgid(&mut self) -> Result<(), String> {
+  fn get_pgid(&self) -> Result<Pid, String> {
     let pid = self.id().ok_or_else(|| "Process had no PID.".to_owned())?;
     let pgid = getpgid(Some(Pid::from_raw(pid as i32)))
       .map_err(|e| format!("Could not get process group id of child process: {}", e))?;
-    // Kill the negative PGID to kill the entire process group.
-    signal::kill(Pid::from_raw(-pgid.as_raw()), signal::Signal::SIGKILL)
+    Ok(pgid)
+  }
+
+  /// Send a signal to the child process group.
+  fn signal_pg<T: Into<Option<signal::Signal>>>(&mut self, signal: T) -> Result<(), String> {
+    let pgid = self.get_pgid()?;
+    // the negative PGID will signal the entire process group.
+    signal::kill(Pid::from_raw(-pgid.as_raw()), signal)
       .map_err(|e| format!("Failed to interrupt child process group: {}", e))?;
+    Ok(())
+  }
+
+  fn has_child_exited(&mut self) -> bool {
+    match self.child.try_wait() {
+      Ok(Some(_)) => true,
+      Ok(None) => false,
+      Err(_) => false,
```

Comment:
This case should probably be handled separately inside `graceful_shutdown`: there is no point waiting longer if we cannot poll the status of the process, so we should probably skip straight to `kill_pgid`.